[Index Management] Add enrich policies fetch api and expose in plugin api

sabarasaba opened the pull request:
Related to https://github.com/elastic/kibana/pull/163159

## Summary

In order to unblock the esql team working on the kibana implementation, I'll first merge the API changes for listing enrich policies and a client side public api interface for exposing this mechanism so that the consumer plugins can use it.

Consumers of this plugin can use public `apiService` from index management by first adding the `indexManagement` plugin as a dependency of their plugin and then proceding to invoke the method as:

```typescript
const { data, error } = await indexManagement?.apiService.getAllEnrichPolicies();
```

Which will then return a serialized list of enriched policies that look something like:

```json
[
    {
      "name": "my-policy",
      "type": "match",
      "sourceIndices": ["users"],
      "matchField": "email",
      "enrichFields": ["first_name", "last_name", "city"]
    },
  ]
```

##### How to test
1. Start elasticsearch with `yarn es snapshot --license=trial` and serverless search solution with `yarn start`
2. Open the [ILM/public/plugin.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/index_lifecycle_management/public/plugin.tsx#L41) file in your text editor and add `indexManagement?.apiService.getAllEnrichPolicies();`
3. Go to console and create a test enrich policy
4. Open ILM  policies page and make sure that a request for fetching `enrich policies` has been fired

<details>
<summary>Create a policy</summary>

```
PUT /users
{
  "mappings": {
    "properties": {
      "email": {
        "type": "keyword"
      },
      "first_name": {
        "type": "text"
      },
      "last_name": {
        "type": "text"
      },
      "city": {
        "type": "text"
      },
      "zip": {
        "type": "keyword"
      },
      "state": {
        "type": "keyword"
      }
    }
  }
}


PUT /_enrich/policy/my-policy
{
  "match": {
    "indices": "users",
    "match_field": "email",
    "enrich_fields": ["first_name", "last_name", "city", "zip", "state"]
  }
}
```
</details>

elasticmachine commented:
Pinging @elastic/platform-deployment-management (Team:Deployment Management)

yuliacech commented on x-pack/plugins/index_management/public/services/public_api_service.ts:
```diff
@@ -0,0 +1,39 @@
+import { API_BASE_PATH } from '../../common/constants';
+
+export interface PublicApiServiceSetup {
+  getAllEnrichPolicies(): void;
```
Should we use the return type here to indicate that this method returns enrich policies?

yuliacech commented on x-pack/plugins/index_management/server/routes/api/enrich_policies/register_list_route.ts:
```diff
@@ -0,0 +1,26 @@
+
+export function registerListRoute({ router, lib: { handleEsError } }: RouteDependencies) {
+  router.get(
+    { path: addBasePath('/enrich_policies'), validate: false },
```
I was thinking that we should consider adding new routes under `/internal/index_management` base path. This would clearly indicate other plugins to now use the route directly and rely on the provided api service instead. It's less effort to do it before the route is published than change it later, wdyt?

sabarasaba commented:
@elasticmachine merge upstream

sabarasaba commented:
Thanks for the review @yuliacech! I've addressed your comments with https://github.com/elastic/kibana/pull/163556/commits/a4e984e540d1aae18dc10a04c7f08da302cd3f3b. In regards of the API integration tests I'll aim to have all of them together in a later iteration to avoid bloating in the PRs. I've created an issue for that and also integrated that into the project plan: https://github.com/elastic/kibana/issues/163807

yuliacech reviewed (approved):
Latest changes LGTM! Thanks a lot for addressing my comments 👍
